python/taichi/lang/matrix.py

```diff
@@ -25,38 +26,33 @@ def __init__(self,
                  keep_raw=False,
                  rows=None,
                  cols=None):
-        # TODO: refactor: use multiple initializer like `ti.Matrix.cols([a, b, c])`
-        # and `ti.Matrix.empty(n, m)` instead of ad-hoc `ti.Matrix(cols=[a, b, c])`.
         self.grad = None
+
         if rows is not None or cols is not None:
+            warnings.warn(
+                f"ti.Matrix(rows=[...]) or ti.Matrix(cols=[...]) is deprecated, use ti.Matrix.rows([...]) or ti.Matrix.cols([...]) instead.",
+                DeprecationWarning,
+                stacklevel=3)
             if rows is not None and cols is not None:
                 raise Exception("cannot specify both rows and columns")
-            if cols is not None:
-                rows = cols
-            self.n = len(rows)
-            if isinstance(rows[0], Matrix):
-                for row in rows:
-                    assert row.m == 1, "inputs must be vectors, ie. size n by 1"
-                    assert row.n == rows[
-                        0].n, "input vectors must be the same shape"
-                self.m = rows[0].n
-                # l-value copy:
-                self.entries = [row(i) for row in rows for i in range(row.n)]
-            elif isinstance(rows[0], list):
-                for row in rows:
-                    assert len(row) == len(
-                        rows[0]), "input lists must be the same shape"
-                self.m = len(rows[0])
-                # l-value copy:
-                self.entries = [x for row in rows for x in row]
-            else:
-                raise Exception(
-                    "rows/cols must be list of lists or lists of vectors")
-            if cols is not None:
-                t = self.transpose()
-                self.n = t.n
-                self.m = t.m
-                self.entries = t.entries
+            self.dt = dt
+            import taichi as ti
+            mat = ti.Matrix.cols(cols) if cols is not None else ti.Matrix.rows(
+                rows)
+            self.n = mat.n
+            self.m = mat.m
+            self.entries = mat.entries
+            return
+
+        elif empty == True:
+            warnings.warn(
+                f"ti.Matrix(n, m, empty=True) is deprecated, use ti.Matrix.empty(n, m) instead",
+                DeprecationWarning,
+                stacklevel=3)
+            self.dt = dt
+            self.entries = [[None] * m for _ in range(n)]
```

Comment:
Will this cause some issue because of the shallow copying in python?

Reply:
What do you mean by shallow copying? Each `None` in entries should be unique, e.g.

```
>>> a = [[None]*4 for _ in range(4)]
>>> a
[[None, None, None, None], [None, None, None, None], [None, None, None, None], [None, None, None, None]]
>>> a[1][1]=2
>>> a
[[None, None, None, None], [None, 2, None, None], [None, None, None, None], [None, None, None, None]]
>>> 
```

Reply:
No worry. None is not a pointer so it's ok.